feat: add -n option to customize package name during initialization

KiranRajeev-KV opened the pull request:
# PR  : feat: add -n option to customize package name during initialization
This PR adds a new `-n` or `--name` option to the `init` command to set a custom name during initialization.
This addresses the issue #50 

### Features Added
- Added `-n, --name` option to the `init` command in `configs.js`.
- Updated `index.js` to get the `name` option and then update `package.json`.

Ashrockzzz2003 commented on bin/index.js:
```diff
@@ -80,17 +89,26 @@ program
   if (!templates[selectedTemplate]) {
-    console.error(chalk.bgRed.white(`Template ${selectedTemplate} does not exist. To see available templates use "qse list".`));
+    console.error(
+      chalk.bgRed.white(
+        `Template ${selectedTemplate} does not exist. To see available templates use "qse list".`
+      )
```
I see that you've worked on improving this log message. Instead of bg-red. Can we have text color to be red. Like highlight the template name that they provided and the error (red) and the `qse list`(yellow).

The bg red kinda is not legible.

![image](https://github.com/user-attachments/assets/1e63e8af-7c84-45bd-8ba5-cee5ea120790)

KiranRajeev-KV replied:
Actually i never edited the log message. i had prettier code formatter locally set up. so it changed the single line code to multiple lines for better readability. that's why you can see theses lines as changed.

But i have made the changes you requested.
It was an easy thing to do.

Ashrockzzz2003 commented on bin/index.js:
```diff
@@ -80,17 +89,26 @@ program
 
 function initCommand(options) {
   const selectedTemplate = options.template || "basic"; // Default to 'basic' if no template is specified
+  const packageName = options.name || "quick-start-express-server"; // Default to 'quick-start-express-server' if no name is specified
```
Add propoer validation for `name`.
Refer npm docs.
https://docs.npmjs.com/cli/v10/configuring-npm/package-json#name

KiranRajeev-KV replied:
I have added package name validation using [validate-npm-package-name](https://www.npmjs.com/package/validate-npm-package-name).

KiranRajeev-KV commented:
@Ashrockzzz2003 Anything else i should add or modify?

Ashrockzzz2003 commented:
Will review @KiranRajeev-KV

Abhinav-ark reviewed (changes requested):
## Output when trying to use with other templates

![image](https://github.com/user-attachments/assets/2c7bff14-54de-4d60-8398-270f3cec4e80)

![image](https://github.com/user-attachments/assets/7599209d-df9e-446c-a716-53ad0883ce79)

![image](https://github.com/user-attachments/assets/3e59f6d5-c1ea-4aad-955f-4d574a722f77)

@KiranRajeev-KV  -n Option not working for templates other than basic.

Ashrockzzz2003 reviewed (changes requested):
@KiranRajeev-KV 
That's great what you've done.

It would be great if you can also do this.

- The dependencies are now hardcoded in `bin/configs.js`. To ease the process of adding a new template, if we can isolate and have a `package.json` for each template and remove the logic from `configs.js`.
- After this, we can also easily maintain the dependency versions inside each template. 
- This will need changes to logic inside of `bin/index.js` too.

The new flow will be like
- Read `package.json` from template folder.
- Apply custom name if provided.
- Install dependencies.
- ... rest are the same.


If you can do this, `bounty` for this task will be changed to `70` @KiranRajeev-KV

KiranRajeev-KV commented:
> @KiranRajeev-KV That's great what you've done.
> 
> It would be great if you can also do this.
> 
> * The dependencies are now hardcoded in `bin/configs.js`. To ease the process of adding a new template, if we can isolate and have a `package.json` for each template and remove the logic from `configs.js`.
> * After this, we can also easily maintain the dependency versions inside each template.
> * This will need changes to logic inside of `bin/index.js` too.
> 
> The new flow will be like
> 
> * Read `package.json` from template folder.
> * Apply custom name if provided.
> * Install dependencies.
> * ... rest are the same.
> 
> If you can do this, `bounty` for this task will be changed to `70` @KiranRajeev-KV

I have implemented the features you requested.

### Changes made
- Removed dependencies for each template from `bin/configs.js`.
- Updated the logic in `bin/index.js` to use the `package.json` from the `template/template_name` directory.
- Added a `package.json` for the basic template (since it was missing).
- Updates tests

Ashrockzzz2003 commented on test/init.test.js:
```diff
@@ -95,6 +96,6 @@ describe('init', () => {
 
     test('invalid template name passed', async () => {
         const { stdout, stderr } = await exec(`node ../../bin/index.js init -t invalid_name`, { cwd: tempDir });
-        expect(stderr).toContain(`Template invalid_name does not exist. To see available templates use "qse list".`);
+        expect(stripAnsi(stderr)).toContain(`Template invalid_name does not exist. To see available templates use "qse list".`);
     });
```
Add more test cases for invalid names and also valid name.

### Invalid cases I can think of

- `>214` chars.
- uppercase.
- include space in name.

### Valid cases I can think of

- `<= 214` chars
- lowercase-only
- any other valid name.

KiranRajeev-KV replied:
I have added the test cases.

Ashrockzzz2003 reviewed (changes requested):
Good work overall @Kiran-Rajeev

Thank you! Please look into the suggested changes.

KiranRajeev-KV commented on bin/index.js:
```diff
@@ -80,32 +90,49 @@ program
+  const templatePath = path.join(__dirname, "../templates", selectedTemplate);
+
   const destinationPath = path.join(targetDir);
   const npmInit = chalk.yellow.bold("npm init");
```
I have removed the unused `npm init`

Ashrockzzz2003 commented on bin/index.js:
```diff
@@ -80,31 +90,40 @@ program
-  const templatePath = path.join(parentDir, "templates", templates[selectedTemplate].name);
-  const destinationPath = path.join(targetDir);
-  const npmInit = chalk.yellow.bold("npm init");
+  const templatePath = path.join(__dirname, "../templates", selectedTemplate);
```
any reason why `../` added now?

Ashrockzzz2003 reviewed (changes requested):
Just a small doubt.

Ashrockzzz2003 reviewed (approved):
LGTM

Abhinav-ark reviewed (approved):
Good Work @KiranRajeev-KV 👏.

Abhinav-ark commented:
Resolves #50.

Abhinav-ark commented:
/bounty 70 @KiranRajeev-KV.
